### PR TITLE
Initial attempt at evil-based key bindings

### PR DESCRIPTION
**Pull request by justbur**
cc @TheBB @syl20bnr

I'm not done with this, but this is at least a starting point. I think some of the choices need some work. I didn't know where to put rename for instance. I also need to fix the popups.

Please give feedback on what you'd like to see changed. Here's a table showing the basic ideas

 Command      | Old | New  
--------------|-----|------
 cherry pick  | a/A |      
 branch       | b   |      
 bisect       | B   |      
 commit       | c   |      
 diff         | d/D |      
 help         | h/? |   
 ediff        | e/E |      
 fetch        | f   |      
 pull         | F   |      
 ignore       | i/I |      
 jump         | j   | g    
 delete       | k   | x    
 untrack      | K   | X    
 log          | l/L |      
 merge        | m   |      
 remote       | M   |      
 next section | n   |  C-j
 submodule    | o   |  C-o    
 prev section | p   | C-k
 push         | P   |      
 rebase       | r   |   
 refresh       | g  | gr
 rename       | R   |    
 stage        | s/S |      
 tag          | t   |      
 notes        | T   |      
 unstage      | u/U |      
 revert       | v/V | o/O
 am           | w   |      
 patch        | W   |      
 reset        | x   | C-r    
 show-refs    | y   |      
 cherry       | Y   |      
 stash        | z/Z |      
 git-cmd      | :   | \|
 run          | !   |      

Other/Additions

 Command                                                | New    
--------------------------------------------------------|--------
 evil-goto-first-line   | gg
 evil-goto-line                                         | G      
 evil-search-next                                       | n      
 evil-search-previous                                   | N      
 set-mark-command                                       | v or V 
 evil-ex                                                | :      
 evil-search-forward                                    | /     
 evil-scroll-up                                             | C-u (if C-u scroll enabled) 
 evil-scroll-page-up                                    | C-b    
 evil-scroll-down                                           | C-d   
 evil-scroll-page-down                                  | C-f    
 magit-section-forward-sibling                          | gj or ]
 magit-section-backward-sibling                   | gk or [
 evil-emacs-state                                       | C-z    
  evil-next-visual-line                                  | j
  evil-previous-visual-line                              | k

### CODE REVIEW

**Comment by justbur** [comment]
@tarsius what's the easiest way to move popup commands?

**Comment by TheBB** [comment]
@justbur `magit-change-popup-key`.

**Comment by TheBB** [comment]
I tried this yesterday. Let me be more specific. :-P

`(magit-change-popup-key 'magit-dispatch-popup :action ?k ?K)`

For example. The popup argument is the symbol of the function that triggers the popup.

**Comment by justbur** [comment]
@TheBB I can give you access to my branch if you want to help :-P

**Comment by TheBB** [comment]
Sure, why not.

**Comment by justbur** [comment]
@TheBB ok you should have access now. I'm just on master. I added a TODO list in the header.

File: evil-magit.el [hunk 1]
```diff
@@ -34,11 +34,302 @@
...
+(evil-define-key 'motion magit-blame-mode-map "q"  'magit-blame-quit)
+(evil-define-key 'motion magit-blame-mode-map "t"  'magit-blame-toggle-headings)
+(evil-define-key 'motion magit-blame-mode-map "\M-w" 'magit-blame-copy-hash)
+
+(evil-define-key 'motion git-rebase-mode-map (kbd "q")    'undefined)
+(evil-define-key 'motion git-rebase-mode-map [remap undo] 'git-rebase-undo)
+(evil-define-key 'motion git-rebase-mode-map (kbd "RET") 'git-rebase-show-commit)
+(evil-define-key 'motion git-rebase-mode-map (kbd "SPC") 'magit-diff-show-or-scroll-up)
+(evil-define-key 'motion git-rebase-mode-map (kbd "!")   'git-rebase-exec) ; was x
+(evil-define-key 'motion git-rebase-mode-map (kbd "c")   'git-rebase-pick)
```
Comment (tarsius) [inline evil-magit.el]: Since `p` isn't used for anything else, you could use it here.

Comment (justbur) [inline evil-magit.el]: Good idea. I was trying to think of a use for p

**Comment by justbur** [comment]
@tarsius For the popups, it seems like `magit-dispatch-popup` might be the only one where the keybindings need to change to be consistent with the bindings in the magit buffers here. Are there any other popups that you know of that have bindings that would need to change?

To give an example, `magit-log-popup` has a binding for `magit-log-head` (bound to `h`), but that's ok with me because there's no other way to reach `magit-log-head` outside of the popup. I'm looking for commands that have one key bound to them in a buffer but another in a popup.

**Comment by justbur** [comment]
@tarsius After a bunch of tweaks, I'm pretty happy with this now (I figured out the popup thing I think). I won't touch it for now if you want to review it or invite others to. Let me know.

**Comment by justbur** [comment]
@mbertheau cc'ing you as another person I know who may provide feedback

**Comment by mbertheau** [comment]
@justbur Thanks for the heads-up. I want to try it but I'm not familiar enough with emacs to know how to install and use this package. I'll try checking out evil-magit in my elpa directory and including it in the package list of the git layer.

From what I can see here I'm very excited!

**Comment by justbur** [comment]
@mbertheau clone the repository and use this in your init file
```elisp
  (with-eval-after-load 'magit
    (load "/path/to/evil-magit/evil-magit.el"))
```

**Comment by mbertheau** [comment]
@justbur Thanks! I can report initial happiness. If anything comes up during the next days I'll comment here. Thanks for the nice work!

**Comment by nixmaniack** [comment]
Seems like I can't do selective staging of lines without reaching to the arrow keys after entering visual selection mode with `V`

`j,k` should work when in visual selection mode for selecting lines for staging/discarding. Current behaviour is, it jumps by hunks. Neither `C-j`, `C-k` work which are bound to visual line selection work as expected until I enter visual selection mode. 

`C-n` `C-p` seem to be conflicting with `evil-mc`. Disabling `evil-mc` didn't help either with `C-n,p` to select visually which seems to be bound to `evil-paste-pop{-next}`.

All above tested on Spacemacs.

**Comment by justbur** [comment]
@nixmaniack fixed in 4f4783f

**Comment by nixmaniack** [comment]
@justbur Works as expected now. Thanks for the quick fix.

**Comment by syl20bnr** [comment]
@justbur Why have you chosen to create dispatch functions that shadow vanilla magit bound functions ?
I moved away from this initial implementation in Spacemacs for evilified buffer to replace it with a pre-command-hook and a submap (https://github.com/syl20bnr/spacemacs/blob/develop/core/core-evilified-state.el#L62-L69) so I wonder what benefits are lost with this change.

**Comment by justbur** [comment]
@syl20bnr I'm not sure one is better than the other. I guess I wasn't thinking of this as a mode or a state with hooks that need to run to activate and deactivate things like pre-command-hooks. Does that make sense?

**Comment by syl20bnr** [comment]
It has not to be a state, any submap could do it, actually the evilified state allows to leverage the evil API for binding definitions but it could have been done manually, the dispatch is still manually done in a custom pre-command-hook.

I believe it is cleaner to have a tiny central place that acts as the dispatcher than replicating the work in every functions but I'm not sure about the side effects of this because I'm not aware of the whole command loop of Emacs, this is why I asked you if you have benefits that we may loose in auto-evilification.

I'm hijacking this PR so we can continue the discussion in the Spacemacs thread if you want.

**Comment by justbur** [comment]
@syl20bnr I'm not sure this is the best implementation. I was more concerned with making good choices for key bindings to start, getting something that works, and then fixing up the implementation later.

**Comment by syl20bnr** [comment]
You are right, make it work first :+1:

**Comment by ralesi** [comment]
Can we include  [ and ]  as bindings as well to navigate sections?

**Comment by justbur** [comment]
Sure why not

**Comment by justbur** [comment]
@ralesi 40c674b

**Comment by cpaulik** [comment]
I find the remapping of `j` and `k` to jump between sections very confusing since now `magit-status` is pretty much the only buffer where I can not use `j` and `k` like I normally would. The same applies for selection in visual mode where this is the only mode where I have to do `v` and then `C-j` and `C-k` instead of the normal keybindings.

**Comment by justbur** [comment]
@cpaulik j and k do what n and p do in the magit bindings. Are you suggesting?
```
C-n -> j
C-p -> k
n -> C-j
p -> C-p
```

**Comment by tarsius** [comment]
> Are there any other popups that you know of that have bindings that would need to change?

Cannot think of any, but for popups which bind the same key to the "default action" which is used to enter the popup, you probably want to keep it that way (i.e. change both bindings).

---

Also sorry for not replying earlier. I just cannot contribute much at this point (also I am taking a break now, so won't be contributing anything for a while). One think I would like to note though, is that I was hoping for something a little simpler. Just saying in case someone else wants to contribute another implementation (whether that be a simpler implementation or one that just takes a different approach).

**Comment by cpaulik** [comment]
Yes, since normal emacs bindings for next and previous line ('C-n' and 'C-p') stay unchanged in magit I would also leave them unchanged for evil. 

Going through sections should be something special like moving by paragraphs. An the `gj` and `gk` bindings are already used in this way.

**Comment by justbur** [comment]
@cpaulik ok makes sense to me

**Comment by vyp** [comment]
Actually I think I agree with @cpaulik too on that one. As in, if it's not done here, I probably have would remapped anyway so that the normal state evil mode `j` and `k` (i.e. move down/up a line) bindings stay consistent with everything else. How does everyone else feel about it?

**Comment by nixmaniack** [comment]
Yes, `jk` would make more sense for current `C-jk`.

**Comment by justbur** [comment]
@tarsius how would you fee about me making this into a separate Melpa package and letting it mature? I'd be happy to give it back in the future

**Comment by justbur** [comment]
I'll flip the jk bindings in a little bit

**Comment by justbur** [comment]
@cpaulik @vyp @nixmaniack I changed the behavior of j and k. Let me know what you think

**Comment by cpaulik** [comment]
I obviously like the `j` and `k` behavior :smile: . Why did you remove the `C-jk` bindings instead of switching them like the commit message says? I think `C-jk` would be good keybindings as an alternative to `g jk` or `[]`

**Comment by justbur** [comment]
@cpaulik I can bring them back. What command do you want them on?

**Comment by cpaulik** [comment]
@justbur I would say `magit-section-forward` and `magit-section-backward`.

**Comment by nixmaniack** [comment]
That's what the commit said actually, Switch role?

**Comment by justbur** [comment]
@cpaulik @nixmaniack The commit message is wrong, sorry (I'll revert it next time I push). Instead I put `gj` and `gk` on `magit-section-forward` and `magit-section-backward` respectively. @cpaulik Do you want `C-j` and `C-k` on those too. There are also commands for jumping further called `magit-section-forward-sibling` and a backwards version. I have `[` and `]` on those now, per @ralesi's suggestion. I'm happy to do whatever you guys want with these, but it would be nice to have a little consensus.

**Comment by nixmaniack** [comment]
I like to use `jk` the same way I use it in other buffers when in normal mode. And `C-jk` is good for `magit-section-{forward,backword}`.

On a side note, I think I had reported similar issue when in status section where I'd expected `hjkl` to work the normal vim way.

**Comment by justbur** [comment]
@nixmaniack Default magit only has movement by line in most buffers, and I follow that here. If I remember right, your previous issue was with not being able to move right and left in the status buffer to copy text.

**Comment by nixmaniack** [comment]
Yes, that's right. It was resolved with `C-jk` which is what was the expected behaviour at that time. But just mentioning as I wanted `jk` there as well, so that we can have consensus on `jk` for visual line movement and have `C-jk` for magit section navigation.

**Comment by cpaulik** [comment]
@justbur I would like `C-jk` to move by sections because it is easier to type than `gj` and `gk` when browsing through several sections. And also easier than `[]` on a :de: keyboard.

**Comment by justbur** [comment]
@cpaulik @nixmaniack You're only answering part of my question. Perhaps I wasn't clear. There are two types of section movement `magit-section-forward` and `magit-section-forward-sibling`. The section usually jumps further. Since I don't have a strong opinion on this, I'd like to know which keys should go to both commands. The table at the beginning of this issue shows the current state. I'd like to come up with better movement commands (based on your recommendation) and make them consistent across all the maps. 

Right now, I here you say `C-jk` should be the shorter `magit-section-forward` and `magit-section-backward`. Would you have any other keys for those commands? What would you put on the longer jumping `magit-section-forward-sibling` and `magit-section-backward-sibling`?

**Comment by nixmaniack** [comment]
This is what is expected right now. 
"j"        evil-next-visual-line
"k"       evil-previous-visual-line
"\C-j" magit-section-forward
"\C-k" magit-section-backward

Since I don't use `magit-section-{forward,backward}-sibling` (that often), but if you ask for suggestion `M-jk` might be the next better spot as in magit they're on `M-np` iirc.

**Comment by justbur** [comment]
I think gj and gk are good fits for the -sibling commands.

**Comment by justbur** [comment]
@cpaulik @nixmaniack pushed the changes. Refer to the table at the beginning of the issue. I think you will both be happy

**Pull request merged**